Enhance helper Pod interface and configuration.

mgoltzsche opened the pull request:
* Make the helper Pod receive only environment variables instead of args (this changes the interface in a non-backward compatible way but is simpler to use and potentially provides more backward compatibility in the future; #175).
* Adds the manager options `--pvc-annotation[-required]` to pass through annotations from the PVC to the PV and to the helper Pod (#164).
* Merge the helper Pod's `data` VolumeMount with the one provided with the template to be able to specify `mountPropagation` within the template (#165).
* Rename `helperPod.yaml` to `helper-pod.yaml` (more convenient and if we break sth we can break this as well).
* Expose `--helper-pod-timeout` option (#178).
* Provide a basic usage example of the new features (`examples/cache`).
* Support forceful termination of the manager binary (2xctrl+c - since this is annoying during development otherwise; #179).

Closes #164.
Closes #165.
Closes #175.
Closes #178.
Relates to #179.
Relates to #127.

mgoltzsche commented on provisioner.go:
```diff
@@ -369,42 +440,48 @@ func (p *LocalPathProvisioner) createHelperPod(action ActionType, cmdsForPath []
 	helperPod.Spec.Volumes = append(helperPod.Spec.Volumes, lpvVolumes...)
-	helperPod.Spec.Containers[0].VolumeMounts = append(helperPod.Spec.Containers[0].VolumeMounts, lpvVolumeMounts...)
-	helperPod.Spec.Containers[0].Command = cmdsForPath
-	helperPod.Spec.Containers[0].Args = []string{"-p", filepath.Join(parentDir, volumeDir),
-		"-s", strconv.FormatInt(sizeInBytes, 10),
-		"-m", volumeMode}
```
This is a breaking change!
Reading these options within a setup/teardown script is pretty cumbersome as the previous examples show.
Therefore I propose to use environment variables for these parameters. Also env vars potentially provide better backward compatibility than options (imagine sb builds a binary using a CLI framework: it would fail by default when new/unknown options are added while, when using env vars, this is not a problem).
However we could leave the options here for now in addition to the env vars, deprecate the options somehow and remove them later. Though eventually we can remove them now since the feature is pretty young anyway and not many people may rely on it / not many systems have to be changed.
WDYT?

@nicktming this affects the changes you introduced in #127.

mgoltzsche commented on main.go:
```diff
@@ -19,26 +21,34 @@ import (
+	FlagConfigMapName         = "configmap-name"
+	EnvConfigMapName          = "CONFIGMAP_NAME"
+	FlagHelperPodFile         = "helper-pod-file"
+	DefaultHelperPodFile      = "helper-pod.yaml"
```
This is another breaking change. I could revert that one but the lower case notation is more convenient and if we break sth here already we might as well break this one now.

mgoltzsche commented:
@yasker please have a look

mgoltzsche commented:
I just rebased this PR and resolved the conflicts.

innobead commented:
Thanks for the contribution. 

It looks you want to solve several issues or enhancements, so I would suggest to seperate this PR first, so we can focus on each problem you are resolving and it will be better for review.

mgoltzsche commented:
@innobead okay, I moved the `mountPropagation` feature into a new PR now: #176 - though I've also added the env var support to it (to reduce merge conflicts within other PRs) but in a backward compatible way.

LarsBingBong commented:
In regards to https://github.com/rancher/local-path-provisioner/pull/177 having been merged. I'm getting the following err:

![image](https://user-images.githubusercontent.com/80959443/159286570-77482df4-4989-4ec8-85e6-a34f67e15336.png)

In the local-path helper `Pod`. I deployed from https://raw.githubusercontent.com/rancher/local-path-provisioner/master/deploy/local-path-storage.yaml as the ReadMe describes - and I did so today.

Sure enough I haven't provided this `VOL_DIR` parameter anywhere. Where am I to do so? I can't seem to find it described anywhere.

Thank you very much.

LarsBingBong commented:
Yeah! That made it work! Thank you very much @derekbit
